### PR TITLE
Refactor new-virtual → create-initial-repo

### LINKED ISSUE #341: New-virtual script fails with error "No such file or directory"
1: Create a new caption-this class repository
- Creating a new class repo fails at line 37 in _new-virtual_ script (No such file or directory) 
` pushd "$REPO_NAME" || return`

- This behavior is due to the line 89 inside the _create-repo_ script which is deleting the folder:
`rm -rf "${REPO_NAME}"`

To have a successful creation of a new class repo the  remove command from create-repo needs to be commented out or additional logic to be implemented. 

If we delete the folder before new-virtual completes we get an error.

### PR DESCRIPTION
**Pull request by parkerbxyz**
Replaces `script/new-virtual` and `script/create-repo` with `script/create-initial-repo`.

**Resolves the following issues:**

- #214
- #218 
- #285 
- #311 
- #321 
- #341 

Includes commits from pull request #334.

### CODE REVIEW

**Comment by AAugustine** [comment]
Tested with a class 19 people. I tried to create github-games repos for the participants.
The script successfully created repos and pull requests for 9 people in the class.
But then the creation of Pull Requests began to fail and returned an error `http: warning: HTTP 403 Forbidden` for each attempt.  I does manage to create repos for all 19 people but fails at creating the pull requests.

Maybe we are hitting a rate limit and could possibly throttle the http requests if that's the case.

I am including a file that contains the output from the command. There was nothing in the `log.out` file.
[refactor-new-virtual-output.txt](https://github.com/githubtraining/training-manual/files/7513085/refactor-new-virtual-output.txt)
e.

**Review by AAugustine — COMMENTED** [review]
I ran into some issues when testing the creating of github-games repos for a class of 19. I left more information in the PR conversation.

**Comment by amyschoen** [comment]
That does appear to be a rate limit @AAugustine.  They do result in a 403, and have two formats.  This is the message for hitting an unauthenticated rate limit:

`API rate limit exceeded for x.x.x.x. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)`

Here's the one for hitting the authenticated rate limit:

`API rate limit exceeded for user ID #.`

**Comment by parkerbxyz** [comment]
[Dealing with secondary rate limits](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits).

> If you're making a large number of `POST`, `PATCH`, `PUT`, or `DELETE` requests for a single user or client ID, wait at least one second between each request.

This should be easy enough to implement.

**Comment by parkerbxyz** [comment]
I added a one second wait before creating issues and PRs via 4ce4461. Let me know if this fixes the issue for you, @AAugustine. 🙂

**Comment by AAugustine** [comment]
Preliminary testing with 8 participants worked well.  I'm going to collect a group of 10-15 people internally to test a larger group since I won't have another large class for about 2 weeks.

**Review by amyschoen — CHANGES_REQUESTED** [review]
This is a no-go for me.

```
Updating contents of README.md... Done.
Updating contents of _config.yml... Done.
usage: http [--json] [--form] [--multipart] [--boundary BOUNDARY] [--compress] [--pretty {all,colors,format,none}] [--style STYLE] [--unsorted]
            [--sorted] [--format-options FORMAT_OPTIONS] [--print WHAT] [--headers] [--body] [--verbose] [--all] [--history-print WHAT] [--stream]
            [--output FILE] [--download] [--continue] [--quiet] [--session SESSION_NAME_OR_PATH | --session-read-only SESSION_NAME_OR_PATH]
            [--auth USER[:PASS]] [--auth-type {basic,digest}] [--ignore-netrc] [--offline] [--proxy PROTOCOL:PROXY_URL] [--follow]
            [--max-redirects MAX_REDIRECTS] [--max-headers MAX_HEADERS] [--timeout SECONDS] [--check-status] [--path-as-is] [--chunked]
            [--verify VERIFY] [--ssl {ssl2.3,tls1,tls1.1,tls1.2}] [--ciphers CIPHERS] [--cert CERT] [--cert-key CERT_KEY] [--ignore-stdin] [--help]
            [--version] [--traceback] [--default-scheme DEFAULT_SCHEME] [--debug]
            [METHOD] URL [REQUEST_ITEM ...]
http: error: unrecognized arguments: --raw
```

Marking it as `Request changes` so it doesn't get merged before I have time to investigate further tomorrow.

File: script/create-initial-repo [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  # Attempt to clone template repo into the temporary directory
+  git clone "$initial_repo_url" "$temp_initial_repo_dir" >>log.out 2>&1 || {
+    print_error "Failed to clone template repository."
+    exit 1
+  }
+}
+
+update_readme() {
+  # Ask for class dates
+  echo "What are the class dates? (e.g. May 4 – 7, 2021)"
```
[the remark this answers is not in the document]
Comment (parkerbxyz) [reply script/create-initial-repo:104]: May the 4th be with you 😉

**Comment by parkerbxyz** [comment]
> This is a no-go for me.

@amyschoen `--raw` was added to HTTPie in [2.5.0](https://github.com/httpie/httpie/releases/tag/2.5.0). Updating should fix this for you. 

If you use Homebrew, this should do the trick:

```
brew upgrade httpie
```

**Comment by amyschoen** [comment]
Thanks @parkerbxyz. I figured it was something like that. I’ll update and test tomorrow once I’m back in front of a computer.

**Comment by amyschoen** [comment]
Haven't gotten a successful class repo created yet.  Still running into issues, but the upgrade fixed the error over raw.

```
Updating contents of README.md... Done.
Updating contents of _config.yml... Done.

http: warning: HTTP 415 Unsupported Media Type

Error! Failed to update branch protection rules.
Creating issue: Add collaborators to this repo... Done.
Creating project: Caption Contest... 
http: warning: HTTP 415 Unsupported Media Type

Done.
Creating project column: To do... 
http: warning: HTTP 404 Not Found

Done.
Creating project column: In progress... 
http: warning: HTTP 404 Not Found

Done.
Creating project column: Done... 
http: warning: HTTP 404 Not Found

Done.
Creating project card: ... 
http: warning: HTTP 404 Not Found


http: warning: HTTP 415 Unsupported Media Type
```

@parkerbxyz, were there any changes to the template repos that I need to bring down to my local versions in GHES?

**Comment by amyschoen** [comment]
Digging in, I'm seeing this:

`--body "Let's see who can create the best captions for our collection of memes."`

Did we change back to memes from Octocats?

**Comment by parkerbxyz** [comment]
> ```
> Error! Failed to update branch protection rules.
> ```

I wonder if the endpoint being used is different on GHES. I’ve only tested on github.com.

> ```
> Creating project: Caption Contest... 
> http: warning: HTTP 415 Unsupported Media Type
> 
> Done.
>
> Creating project column: To do... 
> http: warning: HTTP 404 Not Found
> 
> Done.
> Creating project column: In progress... 
> http: warning: HTTP 404 Not Found
> 
> Done.
> Creating project column: Done... 
> http: warning: HTTP 404 Not Found
> 
> Done.
> ```

If the project fails to be created, the other project-related operations (creating columns and cards) will also fail. We should handle this more gracefully and only try to create columns and cards if the project is successfully created. 

It also looks like the error handling on these functions need a little work. It shouldn’t say “Done” if it failed.

**Comment by amyschoen** [comment]
>  I wonder if the endpoint being used is different on GHES. I’ve only tested on github.com.

I'm starting to suspect it's something like that @parkerbxyz.  I'm troubleshooting and inserting a bunch of random echos to get an idea of where it's going wrong.  Something has changed or is subtly different with the new way of doing things since I had zero issues with using the older version of the scripts when I taught my last class.

**Comment by amyschoen** [comment]
I think it might be headers.  I'm noticing that branch protection has a header of `Accept: application/vnd.github.loki-preview+json` and projects have a header of `Accept: application/vnd.github.inertia-preview+json`.  I'll bet it's in preview in GHES and not in preview on .com.  Testing my theory by adding back the headers in my local version.

**Comment by amyschoen** [comment]
Tricky...you added in required reviews which requires a header of `application/vnd.github.luke-cage-preview+json`.  I got past the issue on branch protection.  I'll add it as a review suggestion and then move on to figuring out the necessary headers for the project calls.

File: script/create-initial-repo [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    "required_status_checks": null,
+    "enforce_admins": true,
+    "required_pull_request_reviews": {
+      "required_approving_review_count": 1
+    },
+    "restrictions": {
+      "users": ["'"$TEACHER"'", "'"$TOKEN_OWNER"'"],
+      "teams": []
+    }
+  }' PUT "$org_repos_endpoint/$initial_repo_name/branches/$default_branch/protection"; then
```
Comment (amyschoen) [inline script/create-initial-repo:159]:
```suggestion
  }' PUT "$org_repos_endpoint/$initial_repo_name/branches/$default_branch/protection" "Accept:application/vnd.github.luke-cage-preview+json"; then
```

File: script/shared_functions [hunk 3]
```diff
@@ -27,6 +27,233 @@ print_done() {
...
+
+  echo -n "Creating project: $name... "
+
+  # Unset $REPO_PROJECT_ID in case it was set previously
+  unset REPO_PROJECT_ID
+
+  # Create the project and get the project ID
+  REPO_PROJECT_ID=$(http --check-status --ignore-stdin --auth \
+    "$TOKEN_OWNER:$TEACHER_PAT" "$projects_endpoint" \
+    name="$name" \
```
Comment (amyschoen) [inline script/shared_functions:101]:
```suggestion
    "$TOKEN_OWNER:$TEACHER_PAT" "$projects_endpoint" \
    "Accept:application/vnd.github.inertia-preview+json" \
    name="$name" \
```

File: script/shared_functions [hunk 4]
```diff
@@ -27,6 +27,233 @@ print_done() {
...
+
+  echo -n "Creating project column: $name... "
+
+  # Unset $COLUMN_ID in case it was set previously
+  unset COLUMN_ID
+
+  # Create the project column and get the column ID
+  COLUMN_ID=$(http --check-status --ignore-stdin --auth \
+    "$TOKEN_OWNER:$TEACHER_PAT" "$project_column_endpoint" \
+    project_id="$project_id" \
```
Comment (amyschoen) [inline script/shared_functions:135]:
```suggestion
    "$TOKEN_OWNER:$TEACHER_PAT" "$project_column_endpoint" \
    "Accept:application/vnd.github.inertia-preview+json" \
    project_id="$project_id" \
```

**Comment by amyschoen** [comment]
Are you sure project cards are working?  The call you're using doesn't match any of the API documentation on either GHES or .com.

https://docs.github.com/en/enterprise-server@3.2/rest/reference/projects#create-a-project-card
https://docs.github.com/en/rest/reference/projects#create-a-project-card

File: script/shared_functions [hunk 5]
```diff
@@ -27,6 +27,233 @@ print_done() {
...
+    shift
+  done
+
+  local project_card_endpoint="https://$INSTANCE_URL/projects/columns/$column_id/cards"
+
+  echo -n "Creating project card: $name... "
+
+  if http --check-status --ignore-stdin --quiet --auth \
+    "$TOKEN_OWNER:$TEACHER_PAT" "$project_card_endpoint" \
+    project_id="$project_id" \
```
Comment (amyschoen) [inline script/shared_functions:166]:
```suggestion
    "$TOKEN_OWNER:$TEACHER_PAT" "$project_card_endpoint" \
    "Accept:application/vnd.github.inertia-preview+json" \
    project_id="$project_id" \
```

**Comment by amyschoen** [comment]
Figured it out.  Cards was fine with the header added even though it doesn't match the API documentation.  @parkerbxyz, can you file a bug with whatever team does the API documentation for that.  Enabling pages was what was failing for me.  I figured out the necessary header.  Once I add that suggestion in, I'm done testing for the week.  Only got through creating the initial class repo.  I'll test everything past that next week.

File: script/create-initial-repo [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  create_issue --repo "$initial_repo_name" --title "Add collaborators to this repo" \
+    --body "$(<"$practice_repos_dir/caption-this/issue_1.md")"
+}
+
+enable_github_pages() {
+  if http --check-status --ignore-stdin --quiet --auth "$TOKEN_OWNER:$TEACHER_PAT" --raw '{
+    "source": {
+      "branch": "'"$default_branch"'"
+    }
+  }' POST "$org_repos_endpoint/$initial_repo_name/pages"; then
```
Comment (amyschoen) [inline script/create-initial-repo:186]:
```suggestion
  }' POST "$org_repos_endpoint/$initial_repo_name/pages" "Accept:application/vnd.github.switcheroo-preview+json"; then
```

**Review by chundric — COMMENTED** [review]
ok

**Comment by parkerbxyz** [comment]
> Figured it out. Cards was fine with the header added even though it doesn't match the API documentation. @parkerbxyz, can you file a bug with whatever team does the API documentation for that. Enabling pages was what was failing for me. I figured out the necessary header.

@amyschoen What version of GHES are you on?

**Comment by parkerbxyz** [comment]
The natural language check is failing because of the following issue: https://github.com/sapegin/textlint-rule-terminology/issues/38.

**Comment by amyschoen** [comment]
> @amyschoen What version of GHES are you on? Maybe you’re on an older version that still requires the preview header? https://github.blog/changelog/2021-10-14-rest-api-preview-promotions/

Looks like we're running 3.1.14.  Does it work to include the headers or does it break newer versions?

**Comment by parkerbxyz** [comment]
> Does it work to include the headers or does it break newer versions?

It seems to work fine with the header on github.com. I don’t have a newer version of GHES readily available to test against, but it should handle the headers just fine. They’re just no longer required in the newer versions.

**Comment by parkerbxyz** [comment]
@amyschoen if you’re good with the changes now that those headers are in there, this PR will auto-merge upon your approving review.

**Comment by amyschoen** [comment]
That works for me. I’ll pull down the current state of the code when I’m back in front of my computer tomorrow and run some quick tests prior to approving.

**Comment by amyschoen** [comment]
@parkerbxyz, I got a little delayed.  I blocked an hour off tomorrow to make sure I get this tested and approved.

**Comment by parkerbxyz** [comment]
No worries. Thanks, @amyschoen!

File: practice-repos/caption-this/issue_1.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+## Comment below to be added as a collaborator
+
+We will all be using this repository today. Even though it's public right now, you will not be able to make changes until you're given the correct permissions. We've automated this process via the GitHub API. Once you comment, we will add you as a collaborator.
+
+You'll start to receive a lot of emails. (❗) You can slow these down immediately by clicking the **Unatch** button at the top of the page and selecting either the **Participating and @mentions** or **Ignore** option.
```
Comment (amyschoen) [inline practice-repos/caption-this/issue_1.md:5]:
```suggestion
You'll start to receive a lot of emails. (❗) You can slow these down immediately by clicking the **Unwatch** button at the top of the page and selecting either the **Participating and @mentions** or **Ignore** option.
```

Comment (amyschoen) [reply practice-repos/caption-this/issue_1.md:5]: @parkerbxyz, just started testing.  Find a minor typo.

**Comment by amyschoen** [comment]
@parkerbxyz, it mostly tests fine.  The initial class repo gets created okay.  Making it private works.  I'm able to get collaborators added fine from the commenters on the initial issue.  Conflict repos and GitHub games are working.

Unfortunately, I'm not getting any slides generated in the initial class repo.  I'm investigating now.

**Comment by amyschoen** [comment]
I commented out the delete.  Looks like the files are successfully created.  Did you mean to have everything prefixed with a 2 instead of sequential numbers like we've used previously?  Names altered to protect our internal user names, but I got 02-Amy.md and 02-Josh.md.  Historically, it would've been one of us with an 02 and one with an 03.

**Comment by amyschoen** [comment]
Looks like branch protection is the issue.

```
{
  "message": "Could not create file: At least 1 approving review is required by reviewers with write access.",
  "documentation_url": "https://docs.github.com/enterprise/3.1/user/articles/about-protected-branches"
}
```

We need to make sure that `Include administrators` is not checked.  It fixed the numbering issue as well.

File: script/create-initial-repo [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  # Add a note to column one
+  create_project_card --project_id "$project_id" --column_id "$column_one" \
+    --note "Each student will update their file in the _slides directory with an image and a caption."
+}
+
+update_branch_protections() {
+  # Update branch protection rules for the default branch
+  if http --check-status --ignore-stdin --quiet --auth "$TOKEN_OWNER:$TEACHER_PAT" --raw '{
+    "required_status_checks": null,
+    "enforce_admins": true,
```
Comment (amyschoen) [inline script/create-initial-repo:151]:
```suggestion
    "enforce_admins": false,
```

**Review by amyschoen — CHANGES_REQUESTED** [review]
@parkerbxyz, I think we should be good after you accept my last couple suggestions.  The first one is a minor typo.  The one I just added on branch protections is a big one.  It means that no slides would be generated for anyone.

File: script/create-initial-repo (hunk 8 above)
Comment (parkerbxyz) [reply script/create-initial-repo:151]: Nice catch. Thanks!

**Review by amyschoen — APPROVED** [review]
Everything is running smoothly now.

**Comment by amyschoen** [comment]
@parkerbxyz, this one's been open for awhile.  Feels good to finally wrap it up.  The new automation is great around the auto merging.

**Comment by parkerbxyz** [comment]
Woohoo! It’s also nice to get 6 issues closed as a result of merging this. 😎 Thanks for your help testing on GHES, @amyschoen! ✨

**Pull request merged**